add with_sequence lookup plugin

jvantuyl opened the pull request:
Plugin allows you to do easy counts for items.  Used as follows:

``` yaml
----
- hosts: all

  tasks:

    # create groups
    - group: name=evens state=present

    - group: name=odds state=present

    # create 20 test users
    - user: name=$item state=present groups=odds
      with_sequence: testuser%02d 1 19 2

    - user: name=$item state=present groups=evens
      with_sequence: testuser%02d 2 20 2
```

You can optionally leave out the stride or format string.

jvantuyl committed:
add with_sequence lookup plugin

Plugin allows you to do easy counts for items.

mpdehaan commented:
Nice.

Seems like we should remove the pattern part, because that pattern value will be $item, and that way there aren't two syntaxes where leaving one off might result in some confusing errors.

Also, easier to remember as it's just like range() in Python then.

Sound good?

jvantuyl commented:
Hmmmmm...  I added pattern because I need zero padded numbers and hexadecimal.  I use it with my provisioning plugin for my cloud provider to create hostnames and such.  It's the difference between:
- host00, host01, host02, ..., host09, host10, host11, host12

and
- host0, host1, host2, ..., host9, host10, host11, host12

For me that's important because some of my scripts need to iterate them in order and we lose the ability to do a natural sort.  This is pretty much an adaptation from the way that I do it in BASH, with [GNU coreutils' seq](http://www.gnu.org/software/coreutils/manual/html_node/seq-invocation.html).

One of the more useful things for me is zero padded hexadecimal with %x.  I use this for MAC address and IPv6 address generation.

I'd be glad to drop the pattern thing, but I still need some mechanism to do zero padding and hex.  Would you feel better if I moved the pattern to the end, where it's less likely to be accidentally used?  Or perhaps switch to args more like "sequence=0:255 format=%02x", or maybe even "start=0 stride=2 end=15 format=%02d" with perhaps a shortcut option like "0-255/2" or something?

mpdehaan commented:
So here is an idea.

Ansible typically does key=value arguments, why not continue this?

with_sequence:  pattern=foo start=X end=Y skip=Z

Basically require the pattern= start= end= and skip=, and then it's obvious
what is what.

Pattern can default to %s

On Wed, Jan 9, 2013 at 2:00 PM, Jayson Vantuyl notifications@github.comwrote:

> Hmmmmm... I added pattern because I need zero padded numbers. I use it
> with my provisioning plugin for my cloud provider to create hostnames and
> such. It's the difference between:
> - host00, host01, host02, ..., host09, host10, host11, host12
> 
> and
> - host0, host1, host2, ..., host9, host10, host11, host12
> 
> For me that's important because some of my scripts need to iterate them in
> order and we lose the ability to do a natural sort.
> 
> This was pretty much an adaptation from the way that I do it in BASH, with GNU
> coreutils' seqhttp://www.gnu.org/software/coreutils/manual/html_node/seq-invocation.html
> .
> 
> Also note that this allows you to do zero padded hexadecimal with %x. I
> also use this for MAC address and IPv6 address generation.
> 
> I'd be glad to drop the pattern thing, but I still need some mechanism to
> do zero padding and hex. Would you feel better if I moved the pattern to
> the end, where it's less likely to be accidentally used? Or perhaps switch
> to args more like "range=0:255 pattern=%02x"?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/1838#issuecomment-12060892.

jvantuyl commented:
That's excellent.  Would you be okay with a shortcut form, too ("0-255/3")?  Basically, it would just some sugar for the most common use case.

mpdehaan commented:
that seems reasonable.

On Wed, Jan 9, 2013 at 2:08 PM, Jayson Vantuyl notifications@github.comwrote:

> That's excellent. Would you be okay with a shortcut form, too ("0-255/3").
> Basically, just some sugar for the most common use case?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/1838#issuecomment-12061223.

jvantuyl commented:
Okay.  I'll work up the parsing changes and we can go from there.

jvantuyl commented:
Okay, I reworked the parsing, improved the error handling, added tests, and documented the heck of out it.

dhozac commented:
Is there any reason to create custom int(x, 0) support rather than using what Python already has? Seems identical to me.

I don't see support for terms being a list. That is part of the lookup plugin contract.

jvantuyl commented:
Ah!  I did not know about int(x,0).  That simplifies things nicely.  I'll strip all of that out.

As for the terms being a list...  Wasn't aware of this.  List of what?  Strings?  Is that documented anywhere?  The lookup_plugin stuff is sparse on the website.

dhozac commented:
List of whatever it is you operate on, so in this case strings. Plugins in general aren't documented much more than in the code. It's something that needs to be done.

dhozac commented:
Also, the first hunk in the documentation is incorrect, as it's talking about 0.8. There are also far more plugins than what it lists. It needs to get its own page or something in the docs.

jvantuyl commented:
Okay, got those done.  Anything else?

dhozac commented:
The list would be of several invocations. So you'd have e.g.

```
with_sequence:
- start=0 end=10
- start=20 end=30
```

with the expected result being [0..10] + [20..30]

jvantuyl commented:
Oh, that seems vastly more useful.  Adapting.

jvantuyl commented:
Random Thought: It would be nice if we could return any iterable (i.e. generator).

Also, is the whole "list of terms" thing implemented, because I can't get it to work in a test playbook...

jvantuyl commented:
Oh, wait.  Typo.  Works fine.

dhozac commented:
Works for me, thanks.
